### ISSUE TITLE
S3 Provider Not Found

### ISSUE THREAD

**Issue by kylerberry** [transloadit/uppy-server#40]
`uppy: Invalid provider options detected. Provider will not be loaded`

Hello, in the [Uppy Server documentation](https://uppy.io/docs/server/#Supported-Providers) s3 is a supported provider, but I do not see it in the src.

**Comment by ifedapoolarewaju** [comment]
Hello @kylerberry, thank you for reporting!

The s3 support is a slightly different one. It doesn't exhibit the same behaviors such as listing files, and uploading files on the server.

Instead the server component of the s3 is used to sign upload requests (signing of request is done on the uppy-server), while the rest of the operations (which is for now only "Upload") are done on the client (i.e uppy-client).

Does this make sense to you?

You can also have a look at [the AwsS3 plugin documentation](https://uppy.io/docs/aws-s3/) which is more detailed on it as it is based mostly on the client side.

**Comment by kylerberry** [comment]
@ifedapoolarewaju that does make sense, but with the above error it doesn't seem I can even sign the request. 

I've done the following on the client:

    uppy.use(AwsS3, {
        host: 'http://localhost:8080'
    })

and set up uppy-server to pass through my existing server, as per the documentation: 

```
app.use(uppy.app({
	providerOptions: {
		s3: {
			key: process.env.AWS_ACCESS_KEY,
			secret: process.env.AWS_SECRET_KEY,
			bucket: process.env.AWS_BUCKET,
			region: process.env.AWS_REGION
		}
	},
	server: {
		host: 'localhost:8080'
	}
}))
```

but the s3 provider is not loaded by uppy-server. Am I implementing this incorrectly?

**Comment by ifedapoolarewaju** [comment]
@kylerberry what happens when you try to sign a request? Since s3 is slightly different from other providers, it doesn't really need to loaded the same way as other providers are.

> uppy: Invalid provider options detected. Provider will not be loaded

This error applies to other providers, not the `s3` provider.
Could you please share how you are trying to sign the request, and what happens when you try it?

**Comment by kylerberry** [comment]
@ifedapoolarewaju Maybe that's my problem, I'm not manually creating a signed request. 

I was under the impression that the uppy-server flow was as follows:

1. uppy frontend client would send any file upload request to 'http://localhost:8080'
2. server-side, `uppy.app()` would return a function used by express `app` that captures the request
3. creates a signed upload request from the credentials provided, and then sends that request to AWS.

I was following the options defined [here](https://uppy.io/docs/server/#Options)

btw thanks for taking the time to help me out.

**Comment by ifedapoolarewaju** [comment]
@kylerberry did you take a look at the s3 doc for uppy client? https://uppy.io/docs/aws-s3/
Basically you would have something like this on the uppy client.

```
uppy.use(AwsS3, {
  host: 'https://url-to-my-uppy-server/'
})
```

and on the server-side, you would instantiate uppy-server with you s3 credentials as you are already doing above.

Is this how you are already using it?

**Comment by kylerberry** [comment]
edit: didn't mean to comment yet. I'm using it in the way you've described above. I'm going to do some more work on it this weekend and see if I make any headway. (didn't mean to close yet)

**Comment by ifedapoolarewaju** [comment]
@kylerberry so what exactly then goes wrong when you try to use it? @goto-bus-stop Might you have any ideas about this?

**Comment by goto-bus-stop** [comment]
I think the `/s3/params` route was accidentally matching the generic Provider routes first, before matching the `/s3` controller. The S3 route was still responding correctly for me and I don't get the warning anymore after flipping the order in which routes are added.

**Comment by goto-bus-stop** [comment]
#42, hope that does the trick for you @kylerberry!

**Comment by kylerberry** [comment]
@ifedapoolarewaju @goto-bus-stop The problem was on my end. I was using express basic-auth auth on my routes. When I excluded the 's3' route it started working as expected. Your comment about uppy's internal routing tipped me off, thanks! Now, how should I go about protecting this route?

@goto-bus-stop I tested both with and without your fix https://github.com/transloadit/uppy-server/pull/42 and both worked as expected, so I think it's save to close that PR.
